RFC: Define End-To-End parameters (#1285)
moq-wg/moq-transport: draft-ietf-moq-transport.md

sequence or exceeds the received MAX_REQUEST_ID, it MUST close the session with
`INVALID_REQUEST_ID`.

## Message Parameters {#params}

Some messages include a Parameters field that encodes optional message elements.
All parameters included in a message MUST be defined in the negotiated version

> of MOQT or negotiated via Setup Parameters.  An endpoint that receives an

Author: Hmm I think I was trying to distinguish "message" params from "setup" params, but might need to be more crisp.

> unknown message parameters MUST close the session with `PROTOCOL_VIOLATION`.

Reviewer: ```suggestion
unknown Parameter MUST close the session with `PROTOCOL_VIOLATION`.
```

are defined in {{setup-params}}. Version-specific parameter types are defined
in {{version-specific-params}}.

### Parameter Scope

> Message Parameters sent by the subscriber are always intended for the peer

Author: I disagree.  NGR is an instruction to the terminating endpoint that could result in another NGR going upstream, but it is not "forwarded" in the same way that objects are.  It's subtle.

> Message Parameters sent by the subscriber are always intended for the peer

Reviewer 2: Alan's way of thinking about this makes sense to me. We likely will have other things with specific aggregation handling.

endpoint and are not forwarded by Relays.  Some message parameters sent by the

> publisher are intended for the peer endpoint and are not forwarded by Relays,

Reviewer: ```suggestion
publisher are scoped to the Session and are not forwarded by Relays,
```

Author: With #1374 I assume _all_ publisher **parameters** are scoped to the session, while "Extension Headers" (eg: properties) are end-to-end.

> while others relate to the Track or subscription and are considered End-to-End,

Author: I agree it will be easier to explain and conceptualize if we don't call everything "parameters" and have this distinction between end-to-end and not.

Objects have expired from cache, their state becomes unknown, and a relay that
handles a downstream request that includes those Objects re-requests them.

> When sent by a publisher, this parameter is End-To-End (see

Reviewer: Nit: This is correct, but given the Publisher is the only one who sends it, how about:
```suggestion
MAX CACHE DURATION is only sent by a publisher, and is End-To-End (see
```

publisher's preference from PUBLISH is used. If omitted from SUBSCRIBE_OK,
TRACK_STATUS_OK, PUBLISH or FETCH, the receiver uses Ascending (0x1).

When sent by a publisher, this parameter is End-To-End (see
{{parameter-scope}}).

> 

Reviewer: I'd still love to get rid of Group Order, but that's TBD.

publisher in SUBSCRIBE_OK or PUBLISH when sending these messages to downstream
subscribers.

> When sent by a publisher, this parameter is End-To-End (see

Reviewer 2: Is there any time this is not sent by a publisher ?

After sending a NEW_GROUP_REQUEST upstream, the request is considered
outstanding until the Largest Group increases.

> When sent by a publisher, this parameter is not End-To-End (see

Reviewer: Isn't this sent by a Subscriber?

{{parameter-scope}}).

> 

Author: ```suggestion
```
